### PR TITLE
[NPUW] Add a pass to convert convolutions to matmul

### PR DESCRIPTION
**Pull request by smirnov-alexey**

### CODE REVIEW

File: src/plugins/intel_npu/src/plugin/npuw/lazy_tensor.cpp [hunk 1]
```diff
@@ -130,6 +130,9 @@ bool LazyTensorImpl::operator==(const LazyTensorImpl& other) const {
     case TransformType::CONVERT:
         // everything is already compared above - skip
         break;
+    case TransformType::RESHAPE:
+        // everything is already compared above - skip
+        break;
```
Comment (dmatveev) [inline src/plugins/intel_npu/src/plugin/npuw/lazy_tensor.cpp:135]:
```suggestion
    case TransformType::THIS:
    case TransformType::CONVERT:
    case TransformType::RESHAPE:
        // everything is already compared above - skip
        break;
```
If clang-format allows such code, ofc.

File: src/plugins/intel_npu/src/plugin/npuw/partitioning/partitioning.cpp [hunk 2]
```diff
@@ -1814,6 +1817,27 @@ void Partitioner::optimize(const std::string& func_name) {
...
+                auto& funcall = func_group.refs[f_idx].get();
+                // Update and move closure according to the newly added parameter
+                // FIXME: rewrite with functional LazyTensors
+                // !!! FIXME !!! not tested yet. Won't work with original params and not consts!
+                funcall._lazy_closure[closure_idx].update(TransformType::RESHAPE, std::monostate{});
+                funcall._lazy_closure.push_back(funcall._lazy_closure[closure_idx]);
+                // Some of the tensors might be in closure - preserve it's 1:1 idx mapping with _lazy_closure
+                funcall._closure.push_back(ov::Tensor());
+            }
+        }
```
Comment (dmatveev) [inline src/plugins/intel_npu/src/plugin/npuw/partitioning/partitioning.cpp:1839]: Are FIXMEs still valid?

Also, I've seen this code before - any chance to generalize it with other such passes?

File: src/plugins/intel_npu/src/plugin/npuw/partitioning/partitioning.cpp [hunk 3]
```diff
@@ -1814,6 +1817,27 @@ void Partitioner::optimize(const std::string& func_name) {
...
+        for (auto&& p : ctx.closures_to_reshape_add) {
+            new_params.push_back(p);
+        }
+        for (auto&& p : ctx.closures_to_reshape_remove) {
+            auto param_idx = f._model->get_parameter_index(p);
+            to_remove.push_back(p);
+            to_remove_idx.insert(param_idx);
+            auto closure_idx = param_idx - f._param_offset;
+            for (std::size_t f_idx = 0; f_idx < func_group.refs.size(); ++f_idx) {
+                auto& funcall = func_group.refs[f_idx].get();
```
Comment (dmatveev) [inline src/plugins/intel_npu/src/plugin/npuw/partitioning/partitioning.cpp:1830]:
```suggestion
            for (auto &&funcall : func_group.refs) {
```
Are you using `f_idx` anywhere?

File: src/plugins/intel_npu/src/plugin/npuw/partitioning/partitioning.cpp [hunk 4]
```diff
@@ -1814,6 +1817,27 @@ void Partitioner::optimize(const std::string& func_name) {
...
+                // Update and move closure according to the newly added parameter
+                // FIXME: rewrite with functional LazyTensors
+                // !!! FIXME !!! not tested yet. Won't work with original params and not consts!
+                funcall._lazy_closure[closure_idx].update(TransformType::RESHAPE, std::monostate{});
```
Comment (dmatveev) [inline src/plugins/intel_npu/src/plugin/npuw/partitioning/partitioning.cpp:1834]: that's not really a reshape as you're not reshaping anything. I'd actually make it a reshape (that is, a generic one - accepting a shape tensor to reshape into), because why not

File: src/plugins/intel_npu/src/plugin/npuw/partitioning/partitioning.cpp [hunk 5]
```diff
@@ -1814,6 +1817,27 @@ void Partitioner::optimize(const std::string& func_name) {
...
+                funcall._lazy_closure.push_back(funcall._lazy_closure[closure_idx]);
```
Comment (dmatveev) [inline src/plugins/intel_npu/src/plugin/npuw/partitioning/partitioning.cpp:1835]: I don't get it to be honest. You reshape a tensor in-place (so it is mutable). Then you push it one more time to the same vector.

What in fact you'd need to do is to push a NEW LazyTensor to a new location.

File: src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.cpp [hunk 6]
```diff
@@ -185,13 +185,88 @@ DQMatMulCWi::DQMatMulCWi() {
...
+// FROM:
+//     ???(Act) ------------------------>
+//     Param(W) -> to(f32) -> Multiply -> MatMul
+//     Param(S) ------------>
+//
+// TO:
+//     ???(Act) -> to(f32) ->
+//     Param(W) -> to(f32) -> MatMul -> Multiply
+//     Param(S) -> Reshape ----------->
+//
```
Comment (dmatveev) [inline src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.cpp:204]: Instead of this, I'd recommend you to extend the existing pattern. You can mark `to_f32` as `optional` there and handle both cases in just one pattern

File: src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.cpp [hunk 7]
```diff
@@ -1461,6 +1536,135 @@ SliceLastMatmulMultiply::SliceLastMatmulMultiply() {
     register_matcher(std::make_shared<opp::Matcher>(res, "SliceLastMatmulMultiply"), std::move(callback));
 }
 
+ConvToMatmul::ConvToMatmul(Context::Ref ctx) {
```
Comment (dmatveev) [inline src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.cpp:1539]: A pic would help here a lot

File: src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.cpp [hunk 8]
```diff
@@ -1461,6 +1536,135 @@ SliceLastMatmulMultiply::SliceLastMatmulMultiply() {
...
+    auto transpose_out = opp::wrap_type<ov::op::v1::Transpose>({conv, opp::any_input()});
+
+    // Note: Use [=] to make sure the above objects stay alive in the callback
+    auto callback = [=](ov::pass::pattern::Matcher& m) {
+        auto& node_to_output = m.get_pattern_value_map();
+
+        auto matched_node_param = node_to_output.at(param).get_node_shared_ptr();
+        auto matched_node_param2 = node_to_output.at(param2).get_node_shared_ptr();
+        auto matched_node_convert = node_to_output.at(convert).get_node_shared_ptr();
+        auto matched_node_tr_input = node_to_output.at(tr_input).get_node_shared_ptr();
```
Comment (dmatveev) [inline src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.cpp:1556]: for things coming from `any_input`, I'd avoid taking pointer to node here. Imagine a node having multiple outputs. You need to deal with a particular output here, not with the whole node.

File: src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.cpp [hunk 9]
```diff
@@ -1461,6 +1536,135 @@ SliceLastMatmulMultiply::SliceLastMatmulMultiply() {
...
+        const auto& shape = matched_param->get_shape();
+        const auto& shape2 = matched_param2->get_shape();
+        const auto& tr_in_shape = matched_node_transpose_in->input(0).get_shape();
+        const auto& tr_out_shape = matched_node_transpose_out->output(0).get_shape();
+
+        if (matched_param->get_element_type() == ov::element::i4 &&
+            matched_param2->get_element_type() == ov::element::f32 && shape.size() == 4 && shape2.size() == 4 &&
+            shape[2] == 1 && shape[3] == 1 && shape2[2] == 1 && shape2[3] == 1 && tr_in_shape.size() == 4 &&
+            tr_out_shape.size() == 4 && tr_in_shape[0] == 1 && tr_in_shape[1] == 1 && tr_out_shape[0] == 1 &&
+            tr_out_shape[1] == 1) {
```
Comment (dmatveev) [inline src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.cpp:1574]: let's add simple mnemonic functions to check these shapes, otherwise it is hard to follow.

What I mean - you check multiple shape vectors here, `shape`, `shape2`,  `tr_in_shape`, `tr_out_shape`. Please write individual checkers for those and tiny lambdas and then call them from your `if`.

File: src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.cpp [hunk 10]
```diff
@@ -1461,6 +1536,135 @@ SliceLastMatmulMultiply::SliceLastMatmulMultiply() {
...
+            }
+            return true;  // root has changed
+        }
+        return false;  // root hasn't changed
+    };
+    register_matcher(std::make_shared<opp::Matcher>(transpose_out, "ConvToMatmul"), std::move(callback));
+}
+
+ConvToMatmul2::ConvToMatmul2(Context::Ref ctx) {
+    auto param = opp::wrap_type<ov::op::v0::Parameter>();
```
Comment (dmatveev) [inline src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.cpp:1604]: All my comments to the above pass apply to this one as well.

File: src/plugins/intel_npu/src/plugin/npuw/util.cpp [hunk 11]
```diff
@@ -367,6 +367,16 @@ ov::Tensor ov::npuw::util::to_f16(const ov::Tensor& t) {
...
 }
 
+ov::Tensor ov::npuw::util::drop_last_dims(const ov::Tensor& t) {
+    const auto& shape = t.get_shape();
+    NPUW_ASSERT(shape.size() == 4);
+    const auto& strides = t.get_strides();
+    return ov::Tensor(t.get_element_type(),
+                      ov::Shape{shape[0], shape[1]},
+                      t.data(),
+                      ov::Strides{strides[0], strides[1]});
```
Comment (dmatveev) [inline src/plugins/intel_npu/src/plugin/npuw/util.cpp:377]: There's a risk of having a Tensor over a dangling pointer, if the result of the previous operation (in the LT unroll pipeline) isn't stored anywhere.

File: src/plugins/intel_npu/src/plugin/npuw/util.hpp [hunk 12]
```diff
@@ -61,6 +61,7 @@ ov::SoPtr<ov::ITensor> view(const ov::SoPtr<ov::ITensor>& src, std::size_t dim,
 void to_f32(const ov::Tensor& in, ov::Tensor& out);
 ov::Tensor to_f16(const ov::Tensor& t);
 ov::Tensor transpose(const ov::Tensor& t);
+ov::Tensor drop_last_dims(const ov::Tensor& t);
```
Comment (dmatveev) [inline src/plugins/intel_npu/src/plugin/npuw/util.hpp:64]: re_sha_pe!

**Review by dmatveev — COMMENTED** [review]
My recommendations here:
1. Please align the patterns according to feedback
2. I will review the LT refactoring to make it a proper baseline here
3. So we align the RESHAPE thing with the rest of the refactored code.

File: src/plugins/intel_npu/src/plugin/npuw/lazy_tensor.cpp (hunk 1 above)
Comment (smirnov-alexey) [reply src/plugins/intel_npu/src/plugin/npuw/lazy_tensor.cpp]: Removed

File: src/plugins/intel_npu/src/plugin/npuw/partitioning/partitioning.cpp (hunk 3 above)
Comment (smirnov-alexey) [reply src/plugins/intel_npu/src/plugin/npuw/partitioning/partitioning.cpp]: Removed

File: src/plugins/intel_npu/src/plugin/npuw/partitioning/partitioning.cpp (hunk 4 above)
Comment (smirnov-alexey) [reply src/plugins/intel_npu/src/plugin/npuw/partitioning/partitioning.cpp]: Removed

File: src/plugins/intel_npu/src/plugin/npuw/partitioning/partitioning.cpp (hunk 5 above)
Comment (smirnov-alexey) [reply src/plugins/intel_npu/src/plugin/npuw/partitioning/partitioning.cpp]: Removed

File: src/plugins/intel_npu/src/plugin/npuw/util.hpp (hunk 12 above)
Comment (smirnov-alexey) [reply src/plugins/intel_npu/src/plugin/npuw/util.hpp]: Re_mov_ed :)

File: src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.cpp (hunk 7 above)
Comment (smirnov-alexey) [reply src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.cpp:1552]: Done

File: src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.cpp (hunk 8 above)
Comment (smirnov-alexey) [reply src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.cpp]: Done

File: src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.cpp (hunk 9 above)
Comment (smirnov-alexey) [reply src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.cpp]: Done

File: src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.cpp (hunk 10 above)
Comment (smirnov-alexey) [reply src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.cpp]: Combined into a single pass

File: src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.cpp (hunk 6 above)
Comment (smirnov-alexey) [reply src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.cpp]: Done

File: src/plugins/intel_npu/src/plugin/npuw/util.hpp [hunk 13]
```diff
@@ -99,6 +99,15 @@ struct Impl {
         return iter->second;
     }
 
+    template <typename K>
+    V& at_or_at_or_at(const K& k1, const K& k2, const K& k3) {
```
Comment (dmatveev) [inline src/plugins/intel_npu/src/plugin/npuw/util.hpp:103]: This may continue forever, seems we need to introduce some monadic wrapper for `at`. :)

Comment (dmatveev) [reply src/plugins/intel_npu/src/plugin/npuw/util.hpp:103]: OR a variadic template could save the day here. BTW.

**Review by dmatveev — APPROVED** [review]
Looks great! Thanks! If only we've had tests on this. 😬
